src/cargo/sources/git/utils.rs

```diff
@@ -453,17 +455,14 @@ impl<'a> GitCheckout<'a> {
                 }
                 new_parent_remote_url.set_path(&new_path);
 
-                match new_parent_remote_url.join(child_url_str) {
-                    Ok(x) => x.to_string(),
-                    Err(err) => Err(err).with_context(|| {
-                        format!(
-                            "failed to parse relative child submodule url `{}` using parent base url `{}`",
-                            child_url_str, new_parent_remote_url
-                        )
-                    })?,
-                }
+                new_parent_remote_url.join(child_url_str).with_context(|| {
+                    format!(
+                        "failed to parse relative child submodule url `{}` using parent base url `{}`",
+                        child_url_str, new_parent_remote_url
+                    )
+                })?
             } else {
-                child_url_str.to_string()
+                Url::parse(child_url_str)?
```

Comment:
I know it is pretty unlikely to happen but maybe it still worth having a decent error message for this. What do you think about something like:
```
failed to parse url for submodule `{child_module_name}`: `{child_url_str}`
```